### DISCUSSION TITLE
How to test components which use `useFormStatus`?

### DISCUSSION THREAD

**Discussion by petewritescode** [vercel/next.js#62962] [Help]
### Summary

Hi, I've built a `SubmitButton` component which uses `useFormStatus` to disable the button while form submission is in progress:

```
import React, { FC, ReactNode } from "react";
import { useFormStatus } from "react-dom";

type SubmitButtonProps = {
  children: ReactNode;
};

export const SubmitButton: FC<SubmitButtonProps> = ({ children }) => {
  const { pending } = useFormStatus();

  return (
    <button type="submit" disabled={pending}>
      {children}
    </button>
  );
};
```

I'm trying to test it using Jest and React Testing Library:

```
import { render, screen } from "@testing-library/react";
import React from "react";
import { SubmitButton } from "./submit-button";

describe("SubmitButton", () => {
  it("renders a submit button", () => {
    render(
      <form>
        <SubmitButton>Submit</SubmitButton>
      </form>
    );

    expect(screen.getByRole("button", { name: "Submit" })).toBeInTheDocument();
  });
});
```

However, when I run the test I get the following error:
```
TypeError: (0 , _reactdom.useFormStatus) is not a function
```

I have the latest version of all packages installed, and have configured Jest to use the `jsdom` environment.

Does anyone know why this is failing?

My only guess is that it's because this is a canary feature. I know Next silently swaps the React version behind the scenes to enable canary features, but perhaps Jest is running the version specified in my `package.json` (`18.2.0`) which doesn't include this? I assumed the version swapping would be taken care of by Next since I'm using `next/jest` in my `jest.config.ts` file.

If anyone can point me in the right direction it'd be greatly appreciated. I've attached a link to a minimal reproduction on CodeSandbox - run `npm run test` to see the error.

Thanks,

Pete

### Example

https://codesandbox.io/p/devbox/jest-issue-86qfjn

**Comment by petewritescode** [comment]
I've confirmed that Next and Jest are using different versions. I rendered `React.version` and `ReactDOM.version` in the component and they both output `18.3.0-canary-2c338b16f-20231116` when running in Next but `18.2.0` when running in Jest.

I thought a fix might be to update my `package.json` to use the canary versions of React and ReactDOM, but this causes lots of peer dependency issues.

**Comment by matheus4lves** [comment]
In short, you have to mock the `useFormStatus` hook.

**Tip:** Add the following script to `package.json`: `"test:watch":"jest --watch"` and run it with `npm run test:watch`.

Then, add this before your test suite (describe block):

```js
jest.mock("react-dom", () => ({
  ...jest.requireActual("react-dom"),
  useFormStatus: jest.fn(),
}))
```

This is a partial mock of the `react-dom` module. Check [Mocking partials](https://jestjs.io/docs/mock-functions#mocking-partials) to learn more about it.

The hook returns a `status` object. Since you're only interested in the `pending` property, you can provide an incomplete `status` object to your test by adding this to your test suite:

```js
beforeEach(() => {
  (useFormStatus as jest.Mock).mockReturnValue({pending: false})
})
```

Working example here: https://codesandbox.io/p/devbox/jest-issue-forked-7f2lnf?workspaceId=ws_E2xE1obNeoL82FXYXntijw

**Bonus:** To mock the true value, add this to your test before rendering the button: `(useFormStatus as jest.Mock).mockReturnValue({pending: true})`.

**Comment by Hagen779** [comment]
Testing Components with useFormStatus
The useFormStatus hook from React must be used within a <form> action context, which makes testing tricky. Here's how to handle it:
The Problem
useFormStatus only works inside form actions and returns { pending: false } by default outside that context.
Solution: Mock the Hook
SubmitButton.tsx (Component to test)

import { useFormStatus } from 'react-dom';

export function SubmitButton() {
  const { pending } = useFormStatus();
  
  return (
    <button type="submit" disabled={pending}>
      {pending ? 'Submitting...' : 'Submit'}
    </button>
  );
}

SubmitButton.test.tsx (Test file)
tsximport { render, screen } from '@testing-library/react';
import { useFormStatus } from 'react-dom';
import { SubmitButton } from './SubmitButton';

// Mock the hook
jest.mock('react-dom', () => ({
  ...jest.requireActual('react-dom'),
  useFormStatus: jest.fn(),
}));

describe('SubmitButton', () => {
  beforeEach(() => {
    // Reset mock before each test
    (useFormStatus as jest.Mock).mockReset();
  });

  it('shows "Submit" when not pending', () => {
    (useFormStatus as jest.Mock).mockReturnValue({ 
      pending: false,
      data: null,
      method: null,
      action: null 
    });

    render(<SubmitButton />);
    expect(screen.getByRole('button')).toHaveTextContent('Submit');
    expect(screen.getByRole('button')).not.toBeDisabled();
  });

  it('shows "Submitting..." when pending', () => {
    (useFormStatus as jest.Mock).mockReturnValue({ 
      pending: true,
      data: null,
      method: null,
      action: null 
    });

    render(<SubmitButton />);
    expect(screen.getByRole('button')).toHaveTextContent('Submitting...');
    expect(screen.getByRole('button')).toBeDisabled();
  });
});
With Vitest
tsximport { describe, it, expect, beforeEach, vi } from 'vitest';
import { render, screen } from '@testing-library/react';
import { useFormStatus } from 'react-dom';
import { SubmitButton } from './SubmitButton';

vi.mock('react-dom', async () => {
  const actual = await vi.importActual('react-dom');
  return {
    ...actual,
    useFormStatus: vi.fn(),
  };
});

describe('SubmitButton', () => {
  beforeEach(() => {
    vi.mocked(useFormStatus).mockReset();
  });

  it('shows "Submit" when not pending', () => {
    vi.mocked(useFormStatus).mockReturnValue({ 
      pending: false,
      data: null,
      method: null,
      action: null 
    });

    render(<SubmitButton />);
    expect(screen.getByRole('button')).toHaveTextContent('Submit');
  });

  it('shows "Submitting..." when pending', () => {
    vi.mocked(useFormStatus).mockReturnValue({ 
      pending: true,
      data: null,
      method: null,
      action: null 
    });

    render(<SubmitButton />);
    expect(screen.getByRole('button')).toHaveTextContent('Submitting...');
  });
});
Testing Form Actions (Integration Test)
tsximport { render, screen, waitFor } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import { LoginForm } from './LoginForm';

it('shows loading state during submission', async () => {
  const mockAction = jest.fn(
    () => new Promise(resolve => setTimeout(resolve, 100))
  );

  render(<LoginForm action={mockAction} />);
  
  const submitButton = screen.getByRole('button', { name: /submit/i });
  
  // Initially not pending
  expect(submitButton).toHaveTextContent('Submit');
  
  // Click submit
  await userEvent.click(submitButton);
  
  // Should show loading state
  expect(submitButton).toHaveTextContent('Submitting...');
  expect(submitButton).toBeDisabled();
  
  // Wait for action to complete
  await waitFor(() => {
    expect(submitButton).toHaveTextContent('Submit');
  });
});

Key Points

Mock at module level - Mock react-dom and specifically override useFormStatus
Return full status object - Include pending, data, method, and action
Reset between tests - Use beforeEach to reset mock state
Test both states - Verify both pending and idle states
Integration tests - For real form behavior, test the entire form component with actual actions
